Predicate filter cannot be used with dynamic resource

### Current and expected behavior

I cannot use a predicate with a dynamic object, due to lack of a `Default` implementation on `kube::api::Object`. The reproducer uses a literal `DynamicObject`, the project I'm working on uses a `kube::api::Object<_, _>` type, but the error is the same (see below for a reproducer).

### Possible solution

I think the `Default` trait bound on the predicate filter resource is actually not required - at least removing it _seems_ to work fine, though I'd like someone more familiar with the codebase to sanity check the idea:

```
impl<St, K, P> Stream for PredicateFilter<St, K, P>
where
    St: Stream<Item = Result<K, Error>>,
    K: Resource,
    K::DynamicType: Eq + Hash,
    P: Predicate<K>,
```

### Additional context

Simple reproducer: 

```rust
use kube::api::{ApiResource, DynamicObject, GroupVersionKind};
use kube::runtime::{
    Controller, PredicateConfig, WatchStreamExt, predicates, reflector, watcher,
};

#[allow(unused)]
fn repro(client: kube::Client) {
    let ar = ApiResource::from_gvk(&GroupVersionKind::gvk("", "v1", "Node"));
    let api = kube::Api::<DynamicObject>::all_with(client, &ar);

    let writer = reflector::store::Writer::new(ar.clone());
    let reader = writer.as_reader();
    let stream = watcher(api, watcher::Config::default())
        .default_backoff()
        .reflect(writer)
        .applied_objects()
        .predicate_filter(predicates::generation, PredicateConfig::default());

    // error[E0277]: the trait bound `ApiResource: Default` is not satisfied
    let controller = Controller::for_stream_with(stream, reader, ar);
}

fn main() {}
```

The error is:

```
error[E0277]: the trait bound `ApiResource: Default` is not satisfied
   --> src/main.rs:30:50
    |
 30 |     let controller = Controller::for_stream_with(stream, reader, ar);
    |                      --------------------------- ^^^^^^ the trait `Default` is not implemented for `ApiResource`
    |                      |
    |                      required by a bound introduced by this call
    |
help: the trait `futures_core::stream::Stream` is implemented for `PredicateFilter<St, K, P>`
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/kube-runtime-4.2.0/src/utils/predicate.rs:189:1
    |
189 | / impl<St, K, P> Stream for PredicateFilter<St, K, P>
190 | | where
191 | |     St: Stream<Item = Result<K, Error>>,
192 | |     K: Resource,
193 | |     K::DynamicType: Default + Eq + Hash,
194 | |     P: Predicate<K>,
    | |____________________^
    = note: required for `PredicateFilter<EventFlatten<Reflect<StreamBackoff<impl Stream<Item = Result<Event<DynamicObject>, Error>> + Send, DefaultBackoff>, DynamicObject>>, ..., ...>` to implement `futures_core::stream::Stream`
note: required by a bound in `Controller::<K>::for_stream_with`
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/kube-runtime-4.2.0/src/controller/mod.rs:830:23
    |
829 |     pub fn for_stream_with(
    |            --------------- required by a bound in this associated function
830 |         trigger: impl Stream<Item = Result<K, watcher::Error>> + Send + 'static,
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Controller::<K>::for_stream_with`
    = note: the full name for the type has been written to '/tmp/predrepro/target/debug/deps/predrepro-d75b381c22d55bb1.long-type-4919667429128668585.txt'
    = note: consider using `--verbose` to print the full type name to the console

For more information about this error, try `rustc --explain E0277`.
error: could not compile `predrepro` (bin "predrepro") due to 1 previous error
```

### Environment

N/A

### Configuration and features

```toml
[dependencies]
kube = { version = "4.2.0", features = ["client", "runtime", "unstable-runtime"] }
k8s-openapi = { version = "0.28", features = ["latest"] }
```

### YAML

None

### Affected crates

kube-runtime

### Would you like to work on fixing this bug?

yes

--- a/kube-runtime/src/utils/predicate.rs
+++ b/kube-runtime/src/utils/predicate.rs
@@ -461,4 +461,60 @@ pub(crate) mod tests {
         let second = filtered.next().now_or_never().unwrap().unwrap().unwrap();
         assert_eq!(second.meta().generation, Some(1));
     }
+
+    #[tokio::test]
+    async fn predicate_filter_compiles_with_dynamic_resource() {
+        use kube::api::DynamicObject;
+
+        let obj: DynamicObject = serde_json::from_value(json!({
+            "apiVersion": "v1",
+            "kind": "Pod",
+            "metadata": {
+                "name": "blog",
+                "generation": Some(1),
+            },
+        }))
+        .unwrap();
+        let data = stream::iter([Ok(obj)]);
+        let mut rx = pin!(PredicateFilter::new(
+            data,
+            predicates::generation,
+            Config::default()
+        ));
+
+        let first = rx.next().now_or_never().unwrap().unwrap().unwrap();
+        assert_eq!(first.meta().generation, Some(1));
+
+        // Mostly, this check is ensuring that predicate filters compile with
+        // `DynamicObject`, not testing any behaviour.
+    }
+
+    #[tokio::test]
+    async fn predicate_filter_compiles_with_custom_type() {
+        use kube::api::Object;
+        use kube_client::api::NotUsed;
+
+        let obj: Object<NotUsed, NotUsed> = serde_json::from_value(json!({
+            "apiVersion": "v1",
+            "kind": "Pod",
+            "metadata": {
+                "name": "blog",
+                "generation": Some(1),
+            },
+            "spec": {},
+        }))
+        .unwrap();
+        let data = stream::iter([Ok(obj)]);
+        let mut rx = pin!(PredicateFilter::new(
+            data,
+            predicates::generation,
+            Config::default()
+        ));
+
+        let first = rx.next().now_or_never().unwrap().unwrap().unwrap();
+        assert_eq!(first.meta().generation, Some(1));
+
+        // Mostly, this check is ensuring that predicate filters compile with
+        // `DynamicObject`, not testing any behaviour.
+    }
 }

--- a/kube-runtime/src/utils/predicate.rs
+++ b/kube-runtime/src/utils/predicate.rs
@@ -488,33 +488,4 @@ pub(crate) mod tests {
         // Mostly, this check is ensuring that predicate filters compile with
         // `DynamicObject`, not testing any behaviour.
     }
-
-    #[tokio::test]
-    async fn predicate_filter_compiles_with_custom_type() {
-        use kube::api::Object;
-        use kube_client::api::NotUsed;
-
-        let obj: Object<NotUsed, NotUsed> = serde_json::from_value(json!({
-            "apiVersion": "v1",
-            "kind": "Pod",
-            "metadata": {
-                "name": "blog",
-                "generation": Some(1),
-            },
-            "spec": {},
-        }))
-        .unwrap();
-        let data = stream::iter([Ok(obj)]);
-        let mut rx = pin!(PredicateFilter::new(
-            data,
-            predicates::generation,
-            Config::default()
-        ));
-
-        let first = rx.next().now_or_never().unwrap().unwrap().unwrap();
-        assert_eq!(first.meta().generation, Some(1));
-
-        // Mostly, this check is ensuring that predicate filters compile with
-        // `DynamicObject`, not testing any behaviour.
-    }
 }

--- a/kube-runtime/src/utils/predicate.rs
+++ b/kube-runtime/src/utils/predicate.rs
@@ -190,7 +190,6 @@ impl<St, K, P> Stream for PredicateFilter<St, K, P>
 where
     St: Stream<Item = Result<K, Error>>,
     K: Resource,
-    K::DynamicType: Eq + Hash,
     P: Predicate<K>,
 {
     type Item = Result<K, Error>;

--- a/kube-runtime/src/utils/predicate.rs
+++ b/kube-runtime/src/utils/predicate.rs
@@ -190,7 +190,7 @@ impl<St, K, P> Stream for PredicateFilter<St, K, P>
 where
     St: Stream<Item = Result<K, Error>>,
     K: Resource,
-    K::DynamicType: Default + Eq + Hash,
+    K::DynamicType: Eq + Hash,
     P: Predicate<K>,
 {
     type Item = Result<K, Error>;

